src/GafferArnold/IECoreArnoldPreview/Renderer.cpp

```diff
@@ -1963,7 +1963,21 @@ class ArnoldLight : public ArnoldObject
 
 			// Update light shader.
 
+			// Destructing the ArnoldShader here will instruct Arnold to destroy AtNodes.
+			// Arnold will not, however, update existing light linking in which this light
+			// might take part. Geometry will still point to an AtNode that we've
+			// handed back to Arnold. When we reconstruct the shader below, we have to
+			// be sure to reuse the exact same AtNode so that linking still works. We
+			// currently ensure this by destroying AtNodes in reversed order of
+			// construction (see ~ArnoldShader).
+			// If Arnold's behaviour regarding how AtNodes are managed were to change,
+			// we'd need to update how we ensure geometry pointing to the correct
+			// node. One option would be a `LightLinkManager` that keeps track of
+			// connections. Note that in most cases we could just change parameters in
+			// place, though, and don't need to reconstruct the shader. An exception would be
+			// when the type of light changes.
```

Comment:
I would revise this a little, because I think "hand back" and "reuse the exact same AtNode" are a bit inaccurate, and suggest that the node isn't actually destroyed. Here's an alternative attempt :

>Delete current light shader, destroying all AtNodes it owns. It is crucial that we do this _before_ constructing a new ArnoldShader (and therefore AtNodes) below, because we are relying on a specific behaviour of the Arnold node allocator. When we destroy the light node, Arnold does not remove it from any of the `light_group` arrays we have assigned to geometry, meaning they will contain a dangling pointer. If we destroy the old AtNode first, we get lucky, and Arnold will allocate the new one at the _exact same address_ as the old one, keeping our arrays valid. We have been accidentally relying on this behaviour for some time, and for now continue to rely on it in lieu of a more complex fix. One subtlety is that when the light shader contains more than one node, we must destroy them in reverse order - see ~ArnoldShader.

>A future fix might involve a `LightLinkManager` that is able to track and patch up any affected light links. Because of the extra bookkeeping involved in such an approach, we would want to keep its use to a minimum. We could achieve that for the common case by editing the light node's parameters in place, only creating a new light node when the type has changed.